src/relax/training/utils.cc

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "utils.h"
+
+namespace tvm {
+namespace relax {
+
+/*! \brief Helper to implement append loss.*/
+class AppendLossMutator : public ExprMutator {
+ public:
+  explicit AppendLossMutator(const SeqExpr& loss_body) : loss_body_(loss_body) {}
+
+  Expr VisitExpr_(const SeqExprNode* seq_expr) override {
+    // mutate only the last block.
+    Array<BindingBlock> blocks;
+    for (int i = 0; i < static_cast<int>(seq_expr->blocks.size()); ++i) {
+      CHECK(seq_expr->blocks[i].as<DataflowBlockNode>())
+          << "All blocks in original functions should be Dataflow Block.";
+      if (i < static_cast<int>(seq_expr->blocks.size()) - 1) {
+        blocks.push_back(seq_expr->blocks[i]);
+      } else {
+        BindingBlock new_block = this->VisitBindingBlock(seq_expr->blocks[i]);
+        if (!new_block->bindings.empty()) {
+          blocks.push_back(new_block);
+        }
+      }
+    }
+    return SeqExpr(blocks, loss_body_->body);
+  }
+
+  BindingBlock VisitBindingBlock_(const DataflowBlockNode* block) override {
```

Comment:
Nit: use `final` if this class is not going to be inherited.
```suggestion
  BindingBlock VisitBindingBlock_(const DataflowBlockNode* block) final {
```